Feature/issue 1213 Trimming boost

rok-cesnovar opened the pull request:
## Summary

The main objective of this PR is to trim the Boost library folder in `/lib`. With this PR we get from `66828 items, totalling 608,4 MB` to `22222 items, totalling 193,9 MB` on the Boost folder. 

This PR also adds the `upgrade-boost.sh` script that can be used to upgrade to newer versions. In this case it was used to "upgrade" from 1.69 to 1.69 and do the trimming/pruning.

It also edits the `upgrade-sundials.sh` with an added -u flag to `git add make/*`. Without the flag this throws an error if the developer has a make/local file (which should be ignored but we try to add it). 

## Tests

/

## Side Effects

Hopefully none.

## Checklist

- [x] Math issue #1213

- [x] Copyright holder: Rok Češnovar

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

rok-cesnovar commented on lib/boost_1.69.0/boost/pending/integer_log2.hpp:
```diff
@@ -4,6 +4,6 @@
 #include <boost/config/header_deprecated.hpp>
 
-// BOOST_HEADER_DEPRECATED("<boost/integer/integer_log2.hpp>");
+BOOST_HEADER_DEPRECATED("<boost/integer/integer_log2.hpp>");
```
@syclik this seems like it was manually commented out. Do we still need to comment this out. Will try to add to the script if so.

rok-cesnovar replied:
Ok, it seems that we do need it.

rok-cesnovar commented:
I manually commented out the `BOOST_HEADER_DEPRECATED` in `pending/integer/integer_log2.hpp`. I decided not to include it in the script since it was fixed for 1.70.0. See discussion [here](https://github.com/boostorg/random/issues/49) and [this](https://github.com/boostorg/random/commit/d217e56535816d6b9d662638674939ef205c57e4) commit. The next upgrade of Boost is going to be to 1.70.0+ so I feel that it does not need to be automated. I could also upgrade do 1.70.0, but since we have no other incentive to do this, I feel that is not necessary.

syclik commented:
I'll take a look soon. Any guidance on what I should look for? Do you know if the build failure is real?

rok-cesnovar commented:
I dont think there is a better way than to go through the 5 commits that are actually readable/reviewable. The other 3 commits are done by the script so they can be reviewed that way. 

The commits and the description of the changes:

https://github.com/stan-dev/math/pull/1223/commits/03a88d769cff896d6a657dbcaec6f8e78f0efa11 adds the -u flag to the upgrade sundials script (explained above in the PR post).

https://github.com/stan-dev/math/pull/1223/commits/14312e2fc68399ef48b9b6b3d17db92ba6bac2cb this one adds the upgrade boost file and will be the largest change to review. Its mostly a copy of the sundials one with the cleanup at the end.

https://github.com/stan-dev/math/pull/1223/commits/c391f4242737988a2cfc4d50b10a4cd1934a0fd6 chmod changes on the upgrade-boost.sh

https://github.com/stan-dev/math/pull/1223/commits/7d37e8b1867a3e73c447fedb25d6edff75edf04d
just adds a check to the script file. If the versions are the same, we dont need to modify the numbers in the makefiles and README. Why even support "upgrading" to the same version? If someone finds some new ways of trimming Boost, they can add the lines in the pruning phase of the script and run it. If you feel this isnt necessary, I can remove it.

https://github.com/stan-dev/math/pull/1223/commits/b5f502a510e5239553aa8644c6a64cd378b8fc39 I went and manually commented out the deprecated header warning in the boost/random. In 1.0.70 this is no longer needed so I decided not to include this in the script but rather add it manually.

Regarding the testing failure:
I am experiencing the same issue that Sebastian fixed in https://github.com/stan-dev/math/pull/1082#issuecomment-447662129 I havent figured out why the `tools/build` subdirectory is not included when using the script. I got busy with other stuff so havent gotten to this yet (I stopped the last Jenkins run on this branch, to no clobber the system cause I knew it would fail) . You can dismiss the review request if you want and I will re-request later when I fix this.

syclik commented:
> 03a88d7 adds the -u flag to the upgrade sundials script (explained above in the PR post).
> 
> It also edits the upgrade-sundials.sh with an added -u flag to git add make/*. Without the flag this throws an error if the developer has a make/local file (which should be ignored but we try to add it).

Awesome. I didn't even realize that would be a problem.

> 14312e2 this one adds the upgrade boost file and will be the largest change to review. Its mostly a copy of the sundials one with the cleanup at the end.

Great. I just looked through that and it looks clean enough for someone else to pick up.

> 7d37e8b
> just adds a check to the script file. If the versions are the same, we dont need to modify the numbers in the makefiles and README. Why even support "upgrading" to the same version? If someone finds some new ways of trimming Boost, they can add the lines in the pruning phase of the script and run it. If you feel this isnt necessary, I can remove it.

That's great! I think that's a perfectly good use case. 

> b5f502a I went and manually commented out the deprecated header warning in the boost/random. In 1.0.70 this is no longer needed so I decided not to include this in the script but rather add it manually.

What's happening here? Do we have a lot of warnings? Do we include the deprecated header directly? Is this a current problem on `develop`?

rok-cesnovar commented:
> What's happening here? Do we have a lot of warnings? Do we include the deprecated header directly? Is this a current problem on `develop`?

We dont have any warnings on develop because this was already commented out on the PR that upgraded to 1.69.0 (see last commit in #1082). Its not included directly but through including boost/random and has been fixed for 1.70.0 (https://github.com/boostorg/random/issues/49).

rok-cesnovar commented:
So the issue is that `lib/boost_*/tools/build/` is in .gitignore, but [the tests need these files](http://d1m1s1b1.stat.columbia.edu:8080/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1223/4/pipeline/24):

```
cd lib/boost_1.69.0; ./bootstrap.sh
./bootstrap.sh: 1: ./bootstrap.sh: ./tools/build/src/engine/build.sh: not found
Building Boost.Build engine with toolset ... 
Failed to build Boost.Build build engine
```

I am not sure how, but @wds15 got this directory through in the last Boost upgrade, but you can see that they were facing the same issue in that PR https://github.com/stan-dev/math/pull/1082#issuecomment-447662129

I am also tagging @seantalts since he added those lines to .gitignore (https://github.com/stan-dev/math/commit/4fbb42788b9fe03ca27c788e972f61e56400014d#diff-a084b794bc0759e7a6b77810e01874f2)

seantalts commented:
Probably that line in the gitignore is too broad and should be narrowed to
just the temporary build files.

On Thu, May 9, 2019 at 15:51 Rok Češnovar <notifications@github.com> wrote:

> So the issue is that lib/boost_*/tools/build/ is in .gitignore, but the
> tests need these files
> <http://d1m1s1b1.stat.columbia.edu:8080/blue/organizations/jenkins/Math%20Pipeline/detail/PR-1223/4/pipeline/24>
> :
>
> cd lib/boost_1.69.0; ./bootstrap.sh
> ./bootstrap.sh: 1: ./bootstrap.sh: ./tools/build/src/engine/build.sh: not found
> Building Boost.Build engine with toolset ...
> Failed to build Boost.Build build engine
>
> I am not sure how, but @wds15 <https://github.com/wds15> got this
> directory through in the last Boost upgrade, but you can see that they were
> facing the same issue in that PR #1082 (comment)
> <https://github.com/stan-dev/math/pull/1082#issuecomment-447662129>
>
> I am also tagging @seantalts <https://github.com/seantalts> since he
> added those lines to .gitignore (4fbb427
> #diff-a084b794bc0759e7a6b77810e01874f2
> <https://github.com/stan-dev/math/commit/4fbb42788b9fe03ca27c788e972f61e56400014d#diff-a084b794bc0759e7a6b77810e01874f2>
> )
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/1223#issuecomment-491041755>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGET3E4EUIHVF6TTX7AUUDPUR6CTANCNFSM4HKWQLQA>
> .
>

rok-cesnovar commented:
I removed the `tools/build/src` from `.gitignore` https://github.com/stan-dev/math/pull/1223/commits/dfae418c3cfb6bbf67d5a59539eb6ee20fa23a4b and added the files that are no longer ignore. Boost now builds.

rok-cesnovar commented:
@scylik this in now ready for another review. See comment above for a short description of changes.
